Save sti files with Pillow

flaviojs opened the pull request:
Was playing around with Pillow and implemented the other things that can be registered.

You can now save RGB and INDEXED sti files with Pillow.
It gives you control of how to save with the parameters to `Image.save`.
You can choose the RGBA spec it uses (defaults to the official spec).
The palette in INDEXED ETRLE images is created from the contents of the images.
ETRLE compression is optional.
You can choose the transparent color (palette index 0).
You can choose how semi-transparent pixels are handled.
Missing image offsets and aux_object_data have default values.

The only thing left is reading the INDEXED images as frames, but I'm not in the mood to play with Pillow anymore so this is it.

PS- what is your versioning policy? I think new capabilities should bump the version

flaviojs committed:
Save sti files with Pillow

selaux commented:
Can you add a handful of tests that do the  write/read  cycle for 8bit and 16bit images and check wether the resulting image is the same. Otherwise this is a lot of code without tests :wink:

flaviojs commented:
Sure.
How do I run the current tests? I'm not familiar with python unit testing.

selaux commented:
Running `nosetests` in the project dir should run the tests.

flaviojs committed:
Fix semi trasnparent color always raising an error.

flaviojs commented:
I installed nosetests and ran it, but got a bunch of errors: [nosetests.log](https://github.com/ja2-stracciatella/ja2-open-toolset/files/3280653/nosetests.log)
The code of those tests look fine, do you have any idea what is going on?

flaviojs commented:
Nevermind, it was python 2 biting me again... after installing python3-nose and python3-mock I was able to run nosetests3 and everything was ok.

Maybe ja2py should assert the python version is 3?

flaviojs committed:
Add ETRLE and raw indexes to the encoder and fix etrle compression.

flaviojs committed:
Add tests for StiImageEncoder.

Rework StiImageEncoder to improve clarity.
ETRLE compress all the cases that I wanted compressed.

flaviojs committed:
Tests for force alpha and force no alpha.

flaviojs committed:
Test and fix small bufsize

flaviojs committed:
Text small bufsize for indexes

flaviojs committed:
Test small bufsize for colors.

flaviojs committed:
Allow bufsize smaller than a pixel.

flaviojs committed:
Test not implemented errors.

flaviojs committed:
Avoid Image._getencoder.

It's not part of the public interface so it can change without notice.

flaviojs committed:
Fix comment

flaviojs committed:
Save and load RGB image.

flaviojs committed:
Rework image save.

Only allow AUX_OBJECT_DATA with ETRLE.
Add roundtrip tests.

flaviojs commented:
I think that's enough tests.

flaviojs commented:
Hmm, I'm not sure about the current color component conversion.
Converting X bits to Y bits:
 * if X > Y, low bits are discarded
 * if X < Y, new low bits are filled with 0

It makes it impossible to get a 0xFF component when converting from the official spec (RGB 565).

selaux commented:
True. Looking at the code a more linear mapping would be nice, e.g. by using floats to scale up.

flaviojs committed:
Produce the same color components values as SDL_GetRGBA

Tests from 0 to 9 bits.

flaviojs commented:
Found [SDL_GetRGBA](https://wiki.libsdl.org/SDL_GetRGBA) and decided to match it's output.

lynxlynxlynx commented:
Is anything still missing here?

flaviojs commented:
This PR is done.

There is stuff that needs to be changed eventually (palette control, subimages as frames, aux data) but I won't work on it in this PR.

lynxlynxlynx commented:
Thanks!
